File: python/functions/transform_to_match.py
"""
Finds the rotation/flip transformation that best aligns the detected grid
with the ideal grid, then applies it to the MRI image.
"""

import numpy as np


def transform_to_match(mri_grid: np.ndarray,
                       grid: np.ndarray,
                       ideal_sideways: np.ndarray) -> np.ndarray:
    """
    Parameters
    ----------
    mri_grid       : 2-D float array – summed MRI image of the phantom grid
    grid           : 2-D float array – binary (NaN/1) detected grid map
    ideal_sideways : 2-D float array – ideal sideways positions (NaN where absent)

    Returns
    -------
    mri_grid_new : 2-D float array – mri_grid after the best-matching transform
    """
    # Build 8 candidate orientations of the grid
    g1 = grid
    g2 = np.rot90(g1)
    g3 = np.rot90(g2)
    g4 = np.rot90(g3)
    g5 = np.fliplr(g1)
    g6 = np.rot90(g5)
    g7 = np.rot90(g6)
    g8 = np.rot90(g7)

    candidates = [g1, g2, g3, g4, g5, g6, g7, g8]

    # Mask of ideal positions (1 where not NaN)
    test = np.where(~np.isnan(ideal_sideways), 1.0, np.nan)

    # Score each orientation: nansum of (candidate + test)
    scores = [np.nansum(c + test) for c in candidates]

    best = int(np.argmax(scores))  # 0-indexed

    transforms = [
        lambda img: img,
        lambda img: np.rot90(img),
        lambda img: np.rot90(np.rot90(img)),
        lambda img: np.rot90(np.rot90(np.rot90(img))),
        lambda img: np.fliplr(img),
        lambda img: np.rot90(np.fliplr(img)),
        lambda img: np.rot90(np.rot90(np.fliplr(img))),
        lambda img: np.rot90(np.rot90(np.rot90(np.fliplr(img)))),
    ]

    return transforms[best](mri_grid)

File: matlab/functions/transformToMatch.m
function [mri_grid_new] = transformToMatch(mri_grid,grid,ideal_sideways)
%UNTITLED2 Summary of this function goes here
%   Detailed explanation goes here

grid1_1 = grid; %grid1_1(~isnan(grid1_1))=1;
grid1_2 = rot90(grid1_1);
grid1_3 = rot90(grid1_2);
grid1_4 = rot90(grid1_3);

grid1_5 = flip(grid1_1,2);
grid1_6 = rot90(grid1_5);
grid1_7 = rot90(grid1_6);
grid1_8 = rot90(grid1_7);

test = ideal_sideways; test(~isnan(test)) = 1; 

test2 = grid1_1 + test; test3(1) = nansum(test2(:));
test2 = grid1_2 + test; test3(2) = nansum(test2(:));
test2 = grid1_3 + test; test3(3) = nansum(test2(:));
test2 = grid1_4 + test; test3(4) = nansum(test2(:));

test2 = grid1_5 + test; test3(5) = nansum(test2(:));
test2 = grid1_6 + test; test3(6) = nansum(test2(:));
test2 = grid1_7 + test; test3(7) = nansum(test2(:));
test2 = grid1_8 + test; test3(8) = nansum(test2(:));

[~,how_to_transform] = max(test3(:));

if how_to_transform == 1
    mri_grid_new = mri_grid;
elseif how_to_transform == 2
    mri_grid_new = rot90(mri_grid);
elseif how_to_transform == 3
    mri_grid_new = rot90(rot90(mri_grid));
elseif how_to_transform == 4
    mri_grid_new = rot90(rot90(rot90(mri_grid)));
elseif how_to_transform == 5
    mri_grid_new = flip(mri_grid,2);
elseif how_to_transform == 6
    mri_grid_new = rot90(flip(mri_grid,2));
elseif how_to_transform == 7
    mri_grid_new = rot90(rot90(flip(mri_grid,2)));
elseif how_to_transform == 8
    mri_grid_new = rot90(rot90(rot90(flip(mri_grid,2))));
end

end